### PR TITLE
add db extensions

### PR DESCRIPTION
**Pull request by mellowagain**
should make it easier to use `libsqlite`.

Current disadvantage is that the passed type `T` for all the extension functions needs to be a `struct`. Primitives, Tuples and Maps are not allowed and will result in a runtime error. We can turn this into a compile error (which id prefer tbh) by adding a new trait bound for called e.g `Sealed` which all `struct`s need to implement and which we simply dont implement for the disallowed types.

this pr still contains `todo!`s bcuz im waiting on #43 for the `DbError`. fixes FP-3787

### CODE REVIEW

**Review by stephlow — APPROVED** [review]

File: fpx/src/main.rs [hunk 1]
```diff
@@ -15,6 +15,7 @@ mod api;
 mod commands;
 pub mod data;
 mod events;
+mod ext;
```
Comment (hatchan) [inline fpx/src/main.rs]: can we move this into data, since it is applicable to the database.

File: fpx/src/ext.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    async fn fetch_optional<T: DeserializeOwned>(&mut self) -> Result<Option<T>>;
+
+    async fn fetch_all<T: DeserializeOwned>(&mut self) -> Result<Vec<T>>;
+}
+
+impl RowsExt for Rows {
+    /// `T` must be a `struct`
+    async fn fetch_one<T: DeserializeOwned>(&mut self) -> Result<T> {
+        match self.next().await? {
+            Some(row) => Ok(de::from_row(&row).context("failed to map into target type")?),
```
Comment (hatchan) [inline fpx/src/ext.rs]: We should also refactor to the new DbError in case this fails.

File: fpx/src/ext.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    async fn fetch_one<T: DeserializeOwned>(&mut self) -> Result<T>;
+
+    async fn fetch_optional<T: DeserializeOwned>(&mut self) -> Result<Option<T>>;
+
+    async fn fetch_all<T: DeserializeOwned>(&mut self) -> Result<Vec<T>>;
+}
+
+impl RowsExt for Rows {
+    /// `T` must be a `struct`
+    async fn fetch_one<T: DeserializeOwned>(&mut self) -> Result<T> {
```
Comment (hatchan) [inline fpx/src/ext.rs]: This fn impl could reuse the `fetch_optional` one and just fail out on a None return value.

File: fpx/src/ext.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            Some(row) => Ok(Some(
+                de::from_row(&row).context("failed to map into target type")?,
+            )),
+            None => Ok(None),
+        }
+    }
+
+    /// `T` must be a `struct`
+    async fn fetch_all<T: DeserializeOwned>(&mut self) -> Result<Vec<T>> {
+        let mut results = Vec::with_capacity(5);
```
Comment (hatchan) [inline fpx/src/ext.rs]: what is up with the capacity 5?

File: fpx/src/data.rs [hunk 5]
```diff
@@ -136,3 +136,63 @@ impl Store {
...
+
+            let fall: Vec<Test> = tx
+                .query("select 1 as test union all select 2 union all select 3", ())
+                .await
+                .unwrap()
+                .fetch_all()
+                .await
+                .unwrap();
+
+            assert_eq!(fone.test, 1);
```
Comment (hatchan) [inline fpx/src/data.rs]: Can we maybe move the asserts next to the fetches?

**Review by hatchan — APPROVED** [review]

File: fpx/src/models.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
+use crate::api::types::Request;
```
Comment (hatchan) [inline fpx/src/models.rs]: We can move this module to the data module and just use Request for the struct name

File: fpx/src/data.rs [hunk 7]
```diff
@@ -142,6 +116,9 @@ pub enum DbError {
     #[error("No rows were returned")]
     NotFound,
 
+    #[error("failed to deserialize into `T`: {message}")]
+    FailedDeserialize { message: String },
+
     #[error("Unexpected number of rows was returned")]
     UnexpectedRowsReturned,
```
Comment (hatchan) [inline fpx/src/data.rs]: Think we can get rid of this variant, right? since we're not checking for more than 1 row in fetch_one?

**Review by hatchan — APPROVED** [review]
One small thing, after that :100:

**Pull request merged**